EventSourcing.Core/Events/EventConverter.cs

```diff
@@ -42,9 +58,31 @@ private class EventType
   public override TEvent Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
   {
     var readerClone = reader;
-    var typeString = JsonSerializer.Deserialize<EventType>(ref readerClone)?.Type;
-    var type = EventTypes[typeString ?? throw new JsonException($"Can't decode Event with type {typeString}")];
+    var typeString = JsonSerializer.Deserialize<EventType>(ref readerClone)?.Type
+      ?? throw new JsonException($"Error while extracting event type string. Does the JSON contain a {nameof(EventType.Type)} field?");
+    var type = EventTypes[typeString];
+      
+    var e = (TEvent) JsonSerializer.Deserialize(ref reader, type);
+    return Migrate(e);
+  }
+  
+  private TEvent Migrate(TEvent e)
+  {
+    while (Migrators.TryGetValue(e.GetType().Name, out var migrator))
+      e = (TEvent) migrator.Convert(e);
 
-    return (TEvent) JsonSerializer.Deserialize(ref reader, type);
+    return e;
+  }
+  
+  private static void ValidateMigrators()
+  {
+    foreach (var (source, m) in Migrators)
+    {
+      var migrator = m;
+      
+      while (Migrators.TryGetValue(migrator.Target.Name, out migrator))
+        if (migrator.Target.Name == source)
+          throw new InvalidOperationException($"Loop detected in event migrators containing {source}");
```

Comment:
I wonder if you should do this recursively ( e.g. going from V1 -> V2 -> V3 -> V1 is a problem)